### PR TITLE
Versioned API

### LINKED ISSUE #36: Versioning
- [ ] Schemas
- [ ] Endpoints
- [ ] App: Project-monitoring-app
- [ ] App: metadata-sync [LATER]

### PR DESCRIPTION
**Pull request by tokland**
Closes #36 

Implementation notes:

- `yarn generate-schemas` now uses play2 servers. Versions generated: 2.30 -> 2.33
- `api.metadata`, `api.dataStore`, `api.models`, and so on, are now lazy and cached, so they are only instantiated when used. That's especially relevant for models, as we have hundreds of them.
- Inference types now have a `D2ApiDefinition` argument where they can get info for a particular API version (for now: schemas and filter type).

Update notes for the users of the API:
- Tested on project-monitoring-app ([PR](https://github.com/EyeSeeTea/project-monitoring-app/pull/219)). Metadata-sync may be trickier to update, let me know if we need some extra features. My approach was simple: create a single file (`src/types/d2-api`) where I re-export things for the Dhis2 version we target. In the rest of the app, we always import from "types/d2-api".
- The old `D2ApiDefault` is now simply `D2Api`.
- Before: `import { D2Api, Id, MetadaPick, D2DataSets } from "d2-api"; `; now: `import { D2Api, Id, MetadaPick, D2DataSets } from "d2-api/2.32"; `
- If you need to connect to an instance, but don't know its version, you can still use `D2ApiBase` (from d2-api). You can perform requests but not access any typed section of the api (only method `getVersion` has been added). We can add more features if necessary, but always having in mind that should be basic features that will work on any Dhis >= 2.30.

### CODE REVIEW

**Comment by SferaDev** [comment]
I've been unable to review it, but since this is a major breaking change, I believe it should be something bigger than 0.2.2

**Comment by tokland** [comment]
Indeed, major version should be increased

**Comment by SferaDev** [comment]
Debugging an issue @xurxodev had while migrating ``metadata-sync`` I found the following inconsistency in ``2.30`` schema.

2.30: https://github.com/EyeSeeTea/d2-api/blob/feature/versioning-36/src/2.30/schemas.ts#L6727
2.32: https://github.com/EyeSeeTea/d2-api/blob/feature/versioning-36/src/2.32/schemas.ts#L6799

The property ``dataElementCategoryCombo`` should be ``categoryCombo``, not sure where is coming this inconsistency since before this PR we should've been targeting ``2.30`` schemas by default and this issue was not present then.

**Comment by tokland** [comment]
Yep, the unversioned d2-api was using a 2.32 schema. Let's compare:

https://play.dhis2.org/2.30/api/schemas/dataElement.json
https://play.dhis2.org/2.32.5/api/schemas/dataElement.json

`fieldName: "dataElementCategoryCombo"` -> `fieldName: "categoryCombo"`

I was using `fieldName` because, at least with 2.32` seemed to get the correct name better than `name`.  That's not the case here. Let's try `const value = property.collectionName || property.name || property.fieldName`. Let me know how it goes. It works for project-monitoring-app (but I am usinga lot less models than you).

Whenever this happens, check the schema/property to deduce how to refine the property name getter.

File: src/utils/cache.ts [hunk 1]
```diff
@@ -1,15 +1,29 @@
-const cacheStore: { [key: string]: any } = {};
+/* Cache a lazy class-property getter */
+export function cached<Res>(_target: any, _key: any, descriptor: TypedPropertyDescriptor<Res>) {
+    let cachedValue: Res | undefined = undefined;
+    const originalGetter = descriptor.get;
+    if (!originalGetter) throw "This decorator can only be applied on class properties";
```
Comment (SferaDev) [inline src/utils/cache.ts:5]:
```suggestion
    if (!originalGetter) throw new Error("This decorator can only be applied on class properties");
```

**Comment by SferaDev** [comment]
Found one issue I believe its root cause is in this PR, yet to be found.

In ``metadata-sync`` we have different instances of ``d2-api``, After initializing any second instance the models internal reference to ``d2-api`` is always the first.

image

This causes some parts of the app to malfunction as it goes to the same instance to fetch data instead of the remote one for example.

**Comment by tokland** [comment]
Caching seems to be bound to the class, will look into it.

**Comment by SferaDev** [comment]
> Caching seems to be bound to the class, will look into it.

cache.ts:

```ts
type ArgumentsCache = Map<string, unknown>;
type FunctionCache = Map<unknown, ArgumentsCache>
type MethodCache = Map<Function, FunctionCache>;

// Cache that stores a map of serialized arguments with their results under function descriptors
const methodCache: MethodCache = new Map();
const functionCache: FunctionCache = new Map();

// Decorator to cache class properties and methods
export const cache = () =>
    function(_target: unknown, _key: string | symbol, descriptor: PropertyDescriptor) {
        const prop = descriptor.value ? "value" : "get";
        const originalFunction = descriptor[prop];
        const map: FunctionCache = new Map();

        descriptor[prop] = function(...args: unknown[]) {
            // Serialize arguments to build a key
            const key = JSON.stringify(args);

            // Create a new map if it's the first time the instance has been cached
            if (!map.has(this)) map.set(this, new Map());
            const cache = map.get(this);

            // If there's a memoized value, return it instead of re-calculating
            const memoizedValue = cache?.get(key);
            if (memoizedValue) return memoizedValue;

            // Compute the default value, store it and return it
            const result = originalFunction.apply(this, args);
            cache?.set(key, result);
            return result;
        };

        methodCache.set(descriptor[prop], map);

        return descriptor;
    };

// Wrapper to memoize functions
export function memoize<Obj extends object | void, Args extends any[], U>(
    fn: (...args: Args) => U
) {
    const map: ArgumentsCache = new Map();

    const result = function(this: Obj, ...args: Args) {
        const key = JSON.stringify(args);
        if (map.has(key)) return map.get(key);

        const result = fn.apply(this, args);
        map.set(key, result);
        return result;
    };

    functionCache.set(result, map);

    return result;
}

// Function to clear memoized storage
export const clear = (fn: Function, instance?: Object) => {
    // Clear method entries
    const methodEntries = methodCache.get(fn);
    if (methodEntries) methodEntries?.get(instance)?.clear();

    // Clear function entries
    const functionEntries = functionCache.get(fn);
    if (functionEntries) functionEntries.clear();
};

// Define a lazy cached property of an object
export function defineLazyCachedProperty<Obj extends object, Key extends keyof Obj, Res>(
    object: Obj,
    name: Key,
    get: () => Res
): void {
    let cachedValue: Res | undefined = undefined;

    Object.defineProperty(object, name, {
        get: () => {
            if (!cachedValue) cachedValue = get();
            return cachedValue;
        },
        enumerable: true,
        configurable: true,
    });
}
```

cache.spec.ts:

```ts
import { cache, clear, memoize } from "../cache";

class TestClass {
    constructor(private multiplier = 1) {}

    public memoized = memoize((x: { value: number }, y: { value: number }) => {
        return Math.random() * this.multiplier + x.value + y.value;
    });

    @cache()
    public getBasic(): number {
        return Math.random() * this.multiplier;
    }

    @cache()
    public getComplex(int: number): number {
        return Math.random() * this.multiplier * int;
    }

    @cache()
    public getMultiple(int1: number, int2: number): number {
        return Math.random() * this.multiplier * int1 * int2;
    }

    public resetBasic(): void {
        clear(this.getBasic, this);
    }

    public resetComplex(): void {
        clear(this.getComplex, this);
    }

    public resetMultiple(): void {
        clear(this.getMultiple, this);
    }

    public resetMemoize(): void {
        clear(this.memoized, this);
    }
}

function baseFunction(int: number) {
    return Math.random() * int;
}

describe("Cache decorator with clearing", () => {
    const test = new TestClass();

    beforeEach(() => {
        test.resetBasic();
        test.resetComplex();
        test.resetMultiple();
        test.resetMemoize();
    });

    it("basic - should be the same number", () => {
        expect(test.getBasic()).toEqual(test.getBasic());
    });

    it("complex - should be the same number", () => {
        expect(test.getComplex(10)).toEqual(test.getComplex(10));
    });

    it("multiple - should be the same number", () => {
        expect(test.getMultiple(10, 20)).toEqual(test.getMultiple(10, 20));
    });

    it("memoize - should be the same number", () => {
        expect(test.memoized({ value: 10 }, { value: 20 })).toEqual(
            test.memoized({ value: 10 }, { value: 20 })
        );
    });

    it("function - should be the same number", () => {
        const memoizedFunction = memoize(baseFunction);
        expect(memoizedFunction(10)).toEqual(memoizedFunction(10));
    });

    it("basic - should be a new number", () => {
        const number = test.getBasic();
        test.resetBasic();
        expect(test.getBasic()).not.toEqual(number);
    });

    it("complex - should be a new number", () => {
        const number = test.getComplex(10);
        test.resetComplex();
        expect(test.getComplex(10)).not.toEqual(number);
    });

    it("multiple - should be a new number", () => {
        const number = test.getMultiple(10, 20);
        test.resetMultiple();
        expect(test.getMultiple(10, 20)).not.toEqual(number);
    });

    it("memoize - should be a new number", () => {
        const number = test.memoized({ value: 10 }, { value: 20 });
        test.resetMemoize();
        expect(test.memoized({ value: 10 }, { value: 20 })).not.toEqual(number);
    });

    it("function - should be a new number", () => {
        const memoizedFunction = memoize(baseFunction);
        const number = memoizedFunction(10);
        clear(memoizedFunction);
        expect(memoizedFunction(10)).not.toEqual(number);
    });

    it("complex - should be a different number", () => {
        expect(test.getComplex(20)).not.toEqual(test.getComplex(10));
    });

    it("multiple - should be a different number", () => {
        expect(test.getMultiple(20, 30)).not.toEqual(test.getMultiple(10, 20));
    });

    it("memoize - should be a different number", () => {
        expect(test.memoized({ value: 10 }, { value: 20 })).not.toEqual(
            test.memoized({ value: 30 }, { value: 20 })
        );
    });

    it("function - should be a different number", () => {
        const memoizedFunction = memoize(baseFunction);
        expect(memoizedFunction(10)).not.toEqual(memoizedFunction(20));
    });

    it("multiple functions - should be a different number", () => {
        const memoizedFunction1 = memoize(baseFunction);
        const memoizedFunction2 = memoize(baseFunction);
        expect(memoizedFunction1(10)).not.toEqual(memoizedFunction2(20));
    });

    it("multiple class instances - should be a different number", () => {
        const test2 = new TestClass(100);
        expect(test.getComplex(10)).not.toEqual(test2.getComplex(10));
    });

    it("different class instances - should be a different number", () => {
        const test1 = new TestClass(100);
        const test2 = new TestClass(100);
        expect(test1.getComplex(10)).not.toEqual(test2.getComplex(10));
    });
});

export {};
```

**Comment by tokland** [comment]
As soon as we are ok this PR (code and features), we should probably merge it, as all new PRs adding extra features will require some merging to be done. We can discuss it at our next team meeting.

**Comment by SferaDev** [comment]
> As soon as we are ok this PR (code and features), we should probably merge it, as all new PRs adding extra features will require some merging to be done. We can discuss it at our next team meeting.

Agreed, we are still not using this code in production for any app yet. But we're actively using in ``development`` for ``metadata-sync``. We still had the bug with the cache, so I'm updating the PR with the code we're using in ``metadata-sync``.

**Comment by adrianq** [comment]
Guys, I am happy to merge it ASAP but two comments:
- @tokland are we using this version in project monitoring? Next app to be released is project monitoring and I just want to confirm this code is compatible and we don't have any problem next week.
- @tokland Can you please test the latest cache implementation @SferaDev has just committed?

**Comment by tokland** [comment]
> * @tokland are we using this version in project monitoring? Next app to be released is project monitoring and I just want to confirm this code is compatible and we don't have any problem next week.

We have an unmerged PR using it: 

https://github.com/EyeSeeTea/project-monitoring-app/pull/219

Actually, we don't need to merge it on this release, I was more worried about the PR here in d2-api.

> * @tokland Can you please test the latest cache implementation @SferaDev has just committed?

Done. Now pointing to the latest release of Alexis, it works fine. We only use one d2Api instance, so only in metadata-sync the cache problem was apparent.

**Comment by adrianq** [comment]
Thanks for this work!

**Comment by adrianq** [comment]
@xurxodev This PR was merged a few days ago and will be part of the next release of d2-api. I think we will be releasing a new version in 2-3 days. I am saying this because this PR breaks compatibility and requires a few modifications on the app side. I am thinking in particular about GEE. If you have any question regarding the required changes feel free to give @tokland a shout! cc @ifoche

**Comment by xurxodev** [comment]
@adrianq thanks for the warning. I know the braking changes because I implemented them in metadata sync. 
I've already talked about this change in gee with nacho for the future.

**Pull request merged**